src/Illuminate/Foundation/Http/Middleware/Precognition.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Illuminate\Foundation\Http\Middleware;
+
+use Closure;
+use Illuminate\Container\Container;
+use Illuminate\Contracts\Routing\ResponseFactory;
+use Illuminate\Http\Response;
+use Illuminate\Routing\Contracts\CallableDispatcher;
+use Illuminate\Routing\Contracts\ControllerDispatcher;
+use Illuminate\Foundation\Routing\PrecognitiveCallableDispatcher;
+use Illuminate\Foundation\Routing\PrecognitiveControllerDispatcher;
+
+class Precognition
+{
+    /**
+     * The container instance.
+     *
+     * @var \Illuminate\Container\Container
+     */
+    protected $container;
+
+    public function __construct(Container $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if (! $this->isAttemptingPrecognition($request)) {
+            return $next($request);
+        }
+
+        $this->prepareForPrecognition($request);
+
+        return $this->withResponse($request, $next($request));
+    }
+
+    /**
+     * Determine if request is attempting to perceive the future.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return bool
+     */
+    protected function isAttemptingPrecognition($request)
+    {
+        return in_array($request->header('Precognition'), ['true', '1'], true);
+    }
+
+    /**
+     * Prepare to handle a precognitive request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @retur void
+     */
+    protected function prepareForPrecognition($request)
+    {
+        $this->container->instance('precognitive', true);
```

Comment:
Instead of needing to set this in the container... could you do something like `$request->attributes->set('precognitive', true)`?

Reply:
Smart. I had forgotten about the request attributes bag and I dig it.

Updated: https://github.com/timacdonald/framework/pull/16/commits/69b53a72188384156a03d576ff66ebadca416b1a